[SLIM] Allow dynamic shape parameters

CharlieFRuan opened the pull request:
This PR adds support for dyanmic shape parameters, specifically substituting llama's appearance of `config.vocab_size` to `llama.vocab_size`. 

### Motivation
Models that share the same architecture may only differ in the vocab size, making us have to recompile a model library. With dynamic vocab size, we can use the same llama model library for llama-2, code-llama, wizardMath, wizardCoder, etc.

### Changes needed
1. Quantization
a. GroupQuantization: During `_dequantize()`, insetad of passing in `IntImm`, ~~pass in `tir.Var` when shape is not int~~ pass in weight's shape, which is a tir.Var (cannot initialize a new var since it would create `vocab_size_1` in script)
 b. No change is needed for AWQ since it does not seem to quantize the embedding layer nor the `lm_head`: https://github.com/mlc-ai/mlc-llm/blob/73c47628fb921eaab2a12ab43275e79d8f26cd60/python/mlc_chat/compiler/quantization/awq_quantization.py#L119-L121

2. `convert_weight.py`
  a. Modify `_check_param()` such that it allows dynamic shape
  b. ~~`_calc_total_params()` skips dynamic-shape parameters~~ Replace `_calc_total_params()` with `loader.stats.total_param_num`

3. Metadata
  a. For the `param` field in metadata, if dynamic, record ~~`["vocab_size", 4096]`~~ `[-1, 4096]`, rather than previously `[32000, 4096]`
b. ~~Skips the param reading~~ Read in `[-1, 4096]` in runtime `cpp/metadata/model.cc`

Depend on https://github.com/apache/tvm/pull/16284

### Performance
Tested on llama-2 on RTX 4090, virtually no performance degradation.
Before:
```
Statistics: ----------- prefill -----------
throughput: 325.278 tok/s
total tokens: 7 tok
total time: 0.022 s
------------ decode ------------
throughput: 183.725 tok/s
total tokens: 256 tok
total time: 1.393 s
```

After:
```
Statistics: ----------- prefill -----------
throughput: 325.510 tok/s
total tokens: 7 tok
total time: 0.022 s
------------ decode ------------
throughput: 185.398 tok/s
total tokens: 256 tok
total time: 1.381 s
```

CharlieFRuan commented:
Hi @junrushao, would really appreciate your suggestions here. Specifically, 2.b and 3.b are not ideal.

For 2.b, I think we could have a `dynamic_shape_map` that gets populated in `_check_shape()`, populating `vocab_size` to either `32000`, `32003`, etc. when comparing `expect` and `actual`. But I am unsure whether it is possible that the quantized weight could have shape like `vocab_size / 4` in the future, a case that I'm unsure what would happen.

For 3.b, I currently do not find the usage of `Param` in `metadata` during runtime. Skipping is definitely not a good option here, but what would be a better way of representing this in runtime? I am not too familiar with tvm runtime, but it seems `ShapeTuple` can only be initialized with concrete values. Perhaps we could populate after loading the actual weight in?

junrushao commented:
# Symbolic Parameters

Hey @CharlieFRuan apologize for the delayed response! I was a bit under the weather lately. It's super cool that you figured out almost everything already, and I'd love to share in a few perspectives to help make it more complete!

C1. First of all, about UX, how about supporting symbolic shape Tensor/Parameters in TVM Relax? Instead of writing:

```python
nn.Embedding(tir.Var("vocab_size", "int64"), config.hidden_size)
```

let's allow a more straightforward syntax:

```python
nn.Embedding("vocab_size", config.hidden_size)
```

This can be done by tweaking `[_tensor_placeholder](https://github.com/apache/tvm/blob/fe89ccc360838847f52eb96181ffe69477000835/python/tvm/relax/frontend/nn/core.py#L684)` adding `str` to a possible element type of `shape`. You will likely have to tweak type annotations of a few `nn.Module`s as well. To make things simpler, we could define:

```python
ShapeVar = Union[int, str, tir.PrimExpr]
```

to semantically enumerate the possible types for shape variables.

C2. Handling of symbolic relationship between dynamic shape tensors. There are actually two issues that could potentially cause problems in the future:

- If we print out TVMScript being exported, it's highly possible that we see two distinct variables for `vocab_size`:

```python
@R.function
def prefill(...):
  vocab_size = T.int64()   # for `embed_tokens`
  vocab_size1 = T.int64()  # for `lm_head`
```

It means the underlying TVM compiler is not aware of symbolic relationships like “your `vocab_size` is the same as mine”.

- two `relax.Function` may share the same `tir.Var`. It is conceptually not a big problem, given each function is a relatively independent component in the `IRModule`, but some passes are just assuming otherwise - I was bitten by this issue a couple of times before. In the example below,

```python
@I.ir_module
class Module:

  @R.function
  def prefill(...):
    vocab_size = T.int64() <== A
    ...

  @R.function
  def decode(...):
    vocab_size = T.int64() <== B
    ...
```

TIR variable A and B share the same physical address in memory.

C3. Representing non-static shapes in MLC’s metadata. Currently, if a parameter’s shape is dynamic, use an `is_dynamic` field to replace `shape` when creating metadata. I don’t have much strong opinion, though I would prefer the following:

```python
"shape": [128, -1]
```

where `-1` means dynamic shape. Note that the two approaches both discard symbolic relationship, but my proposal preserves slightly more data on non-dynamic dimensions.

Let me know what you think!

CharlieFRuan commented:
Hi @junrushao, thank you so much for the detailed response, and I hope you feel better soon!

I am in line with C1 and C3, and can follow up with the changes; I really look forward to contributing to relax!

For C2 on the other hand, what would be a possible solution?

Besides, I might need ~the end of next week to follow up on this if that is fine -- will probably focus on SLM rollout and documentation revamp first (amidst traveling). Thank you!

tqchen commented:
On C2

- First update the nn.Parameter constructor to allow creating a new `tir.Var`, with name "vocab_size"
- Update the relax nn frontend
    -   https://github.com/apache/tvm/blob/unity/python/tvm/relax/frontend/nn/spec.py#L545
    -   Right now notice how we simply pass in a param.shape
    -   Instead we shold do a param.shape remap, that remaps symbolic tir.Var by its name to a new Var when the name first created, and to the same var when we see it in second time.

CharlieFRuan commented:
Makes sense, will follow up. Thank you!

CharlieFRuan commented:
Just updated the PR and description to reflect the changes. I believe the issues we discussed above are all addressed. C1 and C2 (both issues) should be addressed by https://github.com/apache/tvm/pull/16284

tqchen commented:
@CharlieFRuan please rebase and maybe @junrushao can then take another look. This would be important for us to build SLM models to support all Llama variants

junrushao commented:
Hey I'm hyper focused on reviewing the Mixtral PR #1529 right now, and will get back to you guys afterwards

CharlieFRuan commented:
Rebased to conform to https://github.com/mlc-ai/mlc-llm/pull/1566. However, compiling `q4f16_0` runs into error below (`q4f16_1` is fine). Will look into this soon.

```
$ python -m mlc_chat compile dist/Llama-2-7b-q4f16_0/mlc-chat-config.json --device cuda -o dist/Llama-2-7b-q4f16_0-cuda.so
[2024-01-11 06:36:07] INFO auto_config.py:69: Found model configuration: dist/Llama-2-7b-q4f16_0/mlc-chat-config.json
[2024-01-11 06:36:07] INFO auto_device.py:76: Found device: cuda:0
[2024-01-11 06:36:07] INFO auto_device.py:76: Found device: cuda:1
[2024-01-11 06:36:07] INFO auto_target.py:62: Found configuration of target device "cuda:0": {"thread_warp_size": 32, "arch": "sm_89", "max_threads_per_block": 1024, "max_num_threads": 1024, "kind": "cuda", "max_shared_memory_per_block": 49152, "tag": "", "keys": ["cuda", "gpu"]}
[2024-01-11 06:36:07] INFO auto_target.py:94: Found host LLVM triple: x86_64-conda-linux-gnu
[2024-01-11 06:36:07] INFO auto_target.py:95: Found host LLVM CPU: znver2
[2024-01-11 06:36:07] INFO auto_target.py:242: Generating code for CUDA architecture: sm_89
[2024-01-11 06:36:07] INFO auto_target.py:243: To produce multi-arch fatbin, set environment variable MLC_MULTI_ARCH. Example: MLC_MULTI_ARCH=70,72,75,80,86,87,89,90
[2024-01-11 06:36:07] INFO auto_config.py:153: Found model type: llama. Use `--model-type` to override.
Compiling with arguments:
  --config          LlamaConfig(hidden_size=4096, intermediate_size=11008, num_attention_heads=32, num_hidden_layers=32, rms_norm_eps=1e-06, vocab_size=32000, position_embedding_base=10000, context_window_size=4096, prefill_chunk_size=4096, num_key_value_heads=32, head_dim=128, tensor_parallel_shards=1, max_batch_size=80, kwargs={})
  --quantization    GroupQuantize(name='q4f16_0', kind='group-quant', group_size=32, quantize_dtype='int4', storage_dtype='uint32', model_dtype='float16', linear_weight_layout='KN', num_elem_per_storage=8, num_storage_per_group=4, max_int_value=7)
  --model-type      llama
  --target          {"thread_warp_size": 32, "host": {"mtriple": "x86_64-conda-linux-gnu", "tag": "", "kind": "llvm", "mcpu": "znver2", "keys": ["cpu"]}, "arch": "sm_89", "max_threads_per_block": 1024, "max_num_threads": 1024, "kind": "cuda", "max_shared_memory_per_block": 49152, "tag": "", "keys": ["cuda", "gpu"]}
  --opt             flashinfer=1;cublas_gemm=0;cudagraph=0
  --system-lib-prefix ""
  --output          dist/Llama-2-7b-q4f16_0-cuda.so
  --overrides       context_window_size=None;sliding_window_size=None;prefill_chunk_size=None;attention_sink_size=None;max_batch_size=None;tensor_parallel_shards=None
[2024-01-11 06:36:07] INFO compile.py:134: Creating model from: LlamaConfig(hidden_size=4096, intermediate_size=11008, num_attention_heads=32, num_hidden_layers=32, rms_norm_eps=1e-06, vocab_size=32000, position_embedding_base=10000, context_window_size=4096, prefill_chunk_size=4096, num_key_value_heads=32, head_dim=128, tensor_parallel_shards=1, max_batch_size=80, kwargs={})
[2024-01-11 06:36:07] INFO compile.py:153: Exporting the model to TVM Unity compiler
[2024-01-11 06:36:10] INFO compile.py:159: Running optimizations using TVM Unity
[2024-01-11 06:36:10] INFO compile.py:172: Registering metadata: {'model_type': 'llama', 'quantization': 'q4f16_0', 'context_window_size': 4096, 'sliding_window_size': -1, 'attention_sink_size': -1, 'prefill_chunk_size': 4096, 'tensor_parallel_shards': 1, 'kv_cache_bytes': 2147483648}
[2024-01-11 06:36:10] INFO pipeline.py:37: Running TVM Relax graph-level optimizations
[2024-01-11 06:36:12] INFO pipeline.py:37: Lowering to TVM TIR kernels
[2024-01-11 06:36:16] INFO pipeline.py:37: Running TVM TIR-level optimizations
[2024-01-11 06:36:21] INFO pipeline.py:37: Running TVM Dlight low-level optimizations
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/ssd1/cfruan/mlc-llm-repos/mlc-llm-head/python/mlc_chat/__main__.py", line 39, in <module>
    main()
  File "/ssd1/cfruan/mlc-llm-repos/mlc-llm-head/python/mlc_chat/__main__.py", line 24, in main
    cli.main(sys.argv[2:])
  File "/ssd1/cfruan/mlc-llm-repos/mlc-llm-head/python/mlc_chat/cli/compile.py", line 131, in main
    compile(
  File "/ssd1/cfruan/mlc-llm-repos/mlc-llm-head/python/mlc_chat/interface/compile.py", line 224, in compile
    _compile(args, model_config)
  File "/ssd1/cfruan/mlc-llm-repos/mlc-llm-head/python/mlc_chat/interface/compile.py", line 174, in _compile
    args.build_func(
  File "/ssd1/cfruan/mlc-llm-repos/mlc-llm-head/python/mlc_chat/support/auto_target.py", line 215, in build
    relax.build(
  File "/ssd1/cfruan/packages/tvm-unity/python/tvm/relax/vm_build.py", line 335, in build
    mod = pipeline(mod)
          ^^^^^^^^^^^^^
  File "/ssd1/cfruan/packages/tvm-unity/python/tvm/ir/transform.py", line 238, in __call__
    return _ffi_transform_api.RunPass(self, mod)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/ssd1/cfruan/packages/tvm-unity/python/tvm/_ffi/_ctypes/packed_func.py", line 239, in __call__
    raise_last_ffi_error()
  File "/ssd1/cfruan/packages/tvm-unity/python/tvm/_ffi/base.py", line 481, in raise_last_ffi_error
    raise py_err
  File "/ssd1/cfruan/mlc-llm-repos/mlc-llm-head/python/mlc_chat/compiler_pass/pipeline.py", line 138, in _pipeline
    mod = seq(mod)
          ^^^^^^^^
  File "/ssd1/cfruan/packages/tvm-unity/python/tvm/ir/transform.py", line 238, in __call__
    return _ffi_transform_api.RunPass(self, mod)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/ssd1/cfruan/packages/tvm-unity/python/tvm/ir/transform.py", line 307, in _pass_func
    return inst.transform_module(mod, ctx)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/ssd1/cfruan/packages/tvm-unity/python/tvm/dlight/base/transform.py", line 64, in transform_module
    sch = _apply_rules(func, target, self.rules, tunable=False)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/ssd1/cfruan/packages/tvm-unity/python/tvm/dlight/base/transform.py", line 80, in _apply_rules
    space = rule.apply(func, target, tunable)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/ssd1/cfruan/packages/tvm-unity/python/tvm/dlight/gpu/reduction.py", line 104, in apply
    self._sch_inner_spatial(
  File "/ssd1/cfruan/packages/tvm-unity/python/tvm/dlight/gpu/reduction.py", line 296, in _sch_inner_spatial
    tx, _ = sch.split(sch.fuse(*s), factors=[len_tx, None])
                      ^^^^^^^^^^^^
  File "/ssd1/cfruan/packages/tvm-unity/python/tvm/tir/schedule/_type_checker.py", line 340, in wrap
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/ssd1/cfruan/packages/tvm-unity/python/tvm/tir/schedule/schedule.py", line 731, in fuse
    return _ffi_api.ScheduleFuse(self, loops, preserve_unit_iters)  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/ssd1/cfruan/packages/tvm-unity/src/tir/schedule/traced_schedule.cc", line 217, in tvm::tir::TracedScheduleNode::Fuse(tvm::runtime::Array<tvm::tir::LoopRV, void> const&, bool)
    LoopRV result = ConcreteScheduleNode::Fuse(loop_rvs, preserve_unit_loops);
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/ssd1/cfruan/packages/tvm-unity/src/tir/schedule/concrete_schedule.cc", line 387, in tvm::tir::ConcreteScheduleNode::Fuse(tvm::runtime::Array<tvm::tir::LoopRV, void> const&, bool)
    CHECK(!loop_rvs.empty()) << "ValueError: 'fuse' requires at least 1 loop(s)";
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: Traceback (most recent call last):
  1: tvm::tir::TracedScheduleNode::Fuse(tvm::runtime::Array<tvm::tir::LoopRV, void> const&, bool)
        at /ssd1/cfruan/packages/tvm-unity/src/tir/schedule/traced_schedule.cc:217
  0: tvm::tir::ConcreteScheduleNode::Fuse(tvm::runtime::Array<tvm::tir::LoopRV, void> const&, bool)
        at /ssd1/cfruan/packages/tvm-unity/src/tir/schedule/concrete_schedule.cc:387
  File "/ssd1/cfruan/packages/tvm-unity/src/tir/schedule/concrete_schedule.cc", line 387
ValueError: Check failed: (!loop_rvs.empty()) is false: 'fuse' requires at least 1 loop(s)
```

junrushao commented:
Thanks for the PR! This is great. BTW, we will need to update the logics in metadata detection to include dynamic shape parameters, so that the memory estimate could be accurate:

https://github.com/mlc-ai/mlc-llm/blob/main/python/mlc_chat/cli/model_metadata.py#L46-L48

so far, I added a defensive check (all dimensions in a shape be positive) to make sure negative numbers are not counted, but with dynamic shape parameter enabled, it won't give an accurate estimate.

junrushao reviewed (approved):
This PR looks good overall! I left one comment above and let's get it in as soon as it's addressed!

tqchen commented:
Thank you @CharlieFRuan @junrushao ! this is merged, let us followup on more things if needed

Hzfengsy commented:
`q4f16_0` and `q3f16_0` are broken after this PR.

MasterJH5574 commented:
The tensor parallelism support is likely also broken by this PR. When running model with tensor parallelism, we read the model metadata in ahead,
https://github.com/mlc-ai/mlc-llm/blob/33ed13a806d191e694ebe78dfbc4e8204c27cb01/cpp/loader/multi_gpu_loader.cc#L129
and reserve the weight tensor with the shape specified in the metadata on remote workers.
https://github.com/mlc-ai/mlc-llm/blob/33ed13a806d191e694ebe78dfbc4e8204c27cb01/cpp/loader/multi_gpu_loader.cc#L176
https://github.com/mlc-ai/mlc-llm/blob/33ed13a806d191e694ebe78dfbc4e8204c27cb01/cpp/loader/multi_gpu_loader.cc#L115-L116

After this PR, `-1` will be part of the shape when dyn shape parameter exists. And the `NDArray::Empty` taking `-1` in the input shape will lead to error (for example, for Llama2 7B TP=2, it now will invoke `NDArray::Empty((-1, 4096), "float16", device`).

MasterJH5574 commented:
The TP issue should be fixed as long as this TODO get addressed https://github.com/mlc-ai/mlc-llm/blob/33ed13a806d191e694ebe78dfbc4e8204c27cb01/cpp/metadata/json_parser.h#L55-L56

MasterJH5574 commented:
https://github.com/mlc-ai/mlc-llm/pull/1632 This PR should be able to fix.

CharlieFRuan commented:
Thank you so much for the fixes
